File: src/main/java/com/patres/timetable/service/mapper/EntityMapper.java
package com.patres.timetable.service.mapper;

import com.patres.timetable.domain.AbstractApplicationEntity;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public abstract class EntityMapper<EntityType extends AbstractApplicationEntity, EntityDtoType> {

    public List<EntityType> toEntity(List<EntityDtoType> dtoList) {
        if (dtoList == null) {
            return null;
        }

        List<EntityType> list = new ArrayList<>();
        for (EntityDtoType entityDto : dtoList) {
            list.add(toEntity(entityDto));
        }

        return list;
    }

    public List<EntityDtoType> toDto(List<EntityType> entityList) {
        if (entityList == null) {
            return null;
        }

        List<EntityDtoType> list = new ArrayList<>();
        for (EntityType entity : entityList) {
            list.add(toDto(entity));
        }

        return list;
    }


    public Set<EntityDtoType> entitySetToEntityDTOSet(Set<EntityType> entitySet) {
        if (entitySet == null) {
            return null;
        }

        Set<EntityDtoType> entityDtoTypes = new HashSet<>();
        for (EntityType entity : entitySet) {
            entityDtoTypes.add(toDto(entity));
        }

        return entityDtoTypes;
    }

    public Set<EntityType> entityDTOSetToEntitySet(Set<EntityDtoType> entityDtoSet) {
        if (entityDtoSet == null) {
            return null;
        }

        Set<EntityType> entitySet = new HashSet<>();
        for (EntityDtoType entityDto : entityDtoSet) {
            entitySet.add(toEntity(entityDto));
        }

        return entitySet;
    }

    abstract public EntityType toEntity(EntityDtoType entityDto);

    abstract public EntityDtoType toDto(EntityType entity);

    abstract public EntityType fromId(Long id);

}

File: src/main/kotlin/com/patres/timetable/service/mapper/EntityMapper.kt
package com.patres.timetable.service.mapper

import java.util.*

abstract class EntityMapper<EntityType, EntityDtoType> : EntityMapperToDto<EntityType, EntityDtoType>() {

    companion object {
        private const val SHORT_NAME_MAX_LENGTH = 3
        fun getShortName(fullName: String?): String {
            val nameWithoutWhitespaces = fullName?.replace("\\s".toRegex(), "")?: ""
            val maxChar = if (nameWithoutWhitespaces.length < SHORT_NAME_MAX_LENGTH) nameWithoutWhitespaces.length else SHORT_NAME_MAX_LENGTH
            return nameWithoutWhitespaces.substring(0, maxChar).toUpperCase()
        }
    }

    fun toEntity(dtoList: List<EntityDtoType>?): List<EntityType> {
        if (dtoList == null) {
            return ArrayList()
        }

        val list = dtoList.map { toEntity(it) }

        return list
    }

    fun entityDTOSetToEntitySet(entityDtoSet: Set<EntityDtoType>?): Set<EntityType> {
        if (entityDtoSet == null) {
            return HashSet()
        }

        val entitySet = entityDtoSet
            .map { toEntity(it) }
            .toHashSet()

        return entitySet
    }

    abstract fun toEntity(entityDto: EntityDtoType): EntityType

    open fun toDtoWithSampleForm(entity: EntityType): EntityDtoType {
        return toDto(entity)
    }

}
